common/define-grammar.js

```diff
@@ -685,7 +703,19 @@ module.exports = function defineGrammar(dialect) {
 
       tuple_type: $ => choice(
         seq($.readonly, '[', commaSep1($._type), ']'),
-        prec(PREC.TUPLE_TYPE, seq('[', commaSep1($._type), ']'))
+        prec(PREC.TUPLE_TYPE, seq(
+          '[',
+          choice(
+            choice(
+              // Tuple members must all have names or all not have names. (5084)
+              // unlabeled sequence
+              ...tupleTypeContents($._tuple_type_member, $.rest_type),
+              // labeled sequence
+              ...tupleTypeContents($.labeled_tuple_type_member, $.annotated_rest_type)
```

Comment:
For the current implementation

```
src/parser.c
9: #define STATE_COUNT 3206
10: #define LARGE_STATE_COUNT 639
```

Later this week I can open another MR for allowing mixed and comparing both.

Reply:
Opened https://github.com/tree-sitter/tree-sitter-typescript/pull/120. Not much of a difference in the `STATE_COUNT`, but I think I prefer that one since enforcing this rule shouldn't be a parsing concern.

Reply:
Do you agree with closing this in favor of #120 or should I fix https://github.com/tree-sitter/tree-sitter-typescript/pull/118#issuecomment-739153814? The implementation in #120 covers that case already

Reply:
Yeah, let's go with #120.